File: Cpp/20.cpp
//栈
class Solution {
public:
    bool isValid(string s) {
        int n = s.size();
        if(n%2) return false;
        unordered_map <char,char> pairs = {
            {')','('},
            {']','['},
            {'}','{'}
        };
        stack<char> stack;
        for(char ch : s) {
            if(pairs.count(ch)) {
                if(stack.empty() || stack.top()!=pairs[ch]) {
                    return false;
                }
                stack.pop();
            } else {
                stack.push(ch);
            }
        } 
        return stack.empty();
    }
};

File: JavaScript/20.js
var isValid = function(s) {
    let map = new Map([
        [')', '('],
        ['}', '{'],
        [']', '['],
    ]);
    let stack = [];
    for(let i=0; i<s.length; ++i) {
        if(s[i] === '(' || s[i] === '{' || s[i] === '[') {
            stack.push(s[i]);
        } else if(map.get(s[i]) === stack[stack.length-1]) {
            stack.pop();
        } else {
            return false;
        }
    }
    return stack.length === 0;
};